src/Data/Component.elm

```diff
@@ -0,0 +1,328 @@
+module Data.Component exposing
+    ( Amount
+    , Component
+    , ComponentItem
+    , Id
+    , ProcessItem
+    , Quantity
+    , Results
+    , amountToFloat
+    , available
+    , componentItemToString
+    , compute
+    , decodeComponentItem
+    , decodeList
+    , emptyResults
+    , encodeComponentItem
+    , expandComponentItems
+    , expandProcessItems
+    , extractImpacts
+    , extractItems
+    , extractMass
+    , findById
+    , idFromString
+    , idToString
+    , quantityFromInt
+    , quantityToInt
+    )
+
+import Data.Impact as Impact exposing (Impacts)
+import Data.Process as Process exposing (Process)
+import Data.Uuid as Uuid exposing (Uuid)
+import Json.Decode as Decode exposing (Decoder)
+import Json.Decode.Pipeline as Decode
+import Json.Encode as Encode
+import Mass exposing (Mass)
+import Quantity
+import Result.Extra as RE
+
+
+type Id
+    = Id Uuid
+
+
+{-| A Component is a named collection of processes and amounts of them
+-}
+type alias Component =
+    { id : Id
+    , name : String
+    , processes : List ProcessItem
+    }
+
+
+{-| A compact representation of a component and a quantity of it
+-}
+type alias ComponentItem =
+    { id : Id
+    , quantity : Quantity
+    }
+
+
+{-| A Db-like interface holding components and processes
+-}
+type alias DataContainer db =
+    { db
+        | components : List Component
+        , processes : List Process
+    }
+
+
+{-| A compact representation of a component process and an amount of it
+-}
+type alias ProcessItem =
+    { amount : Amount
+    , processId : Process.Id
+    }
+
+
+type Amount
+    = Amount Float
+
+
+type Quantity
+    = Quantity Int
+
+
+{-| A nested data structure carrying the impacts and mass resulting from a computation
+-}
+type Results
+    = Results
+        { impacts : Impacts
+        , items : List Results
+        , mass : Mass
+        }
+
+
+{-| Add two results together
+-}
+addResults : Results -> Results -> Results
+addResults (Results results) (Results acc) =
+    Results
+        { acc
+            | impacts = Impact.sumImpacts [ results.impacts, acc.impacts ]
+            , items = Results results :: acc.items
+            , mass = Quantity.sum [ results.mass, acc.mass ]
+        }
+
+
+amountToFloat : Amount -> Float
+amountToFloat (Amount float) =
+    float
+
+
+{-| List components which ids are not part of the provided list of ids
+-}
+available : List Id -> List Component -> List Component
+available ids =
```

Comment:
Even if the comment helps, I would have named the parameter `usedIds` or `excludedIds`, something more explicit than just `ids`.